src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/net/GRPCConnectionManager.java

```diff
@@ -43,24 +47,37 @@
   private static final Logger LOG = LogManager.getLogger(GRPCConnectionManager.class);
   private static final String EMPTY_STRING = "";
 
-  private Map<String, ManagedChannel> perHostChannelMap = new HashMap<>();
-  private Map<String, InterNodeRpcServiceGrpc.InterNodeRpcServiceStub> perHostClientStubMap =
-      new HashMap<>();
+  private ConcurrentMap<String, AtomicReference<ManagedChannel>> perHostChannelMap =
+      new ConcurrentHashMap<>();
+  private ConcurrentMap<String, AtomicReference<InterNodeRpcServiceStub>> perHostClientStubMap =
+      new ConcurrentHashMap<>();
 
   private final boolean shouldUseHttps;
 
   public GRPCConnectionManager(final boolean shouldUseHttps) {
     this.shouldUseHttps = shouldUseHttps;
   }
 
-  public InterNodeRpcServiceGrpc.InterNodeRpcServiceStub getClientStubForHost(
+  public InterNodeRpcServiceStub getClientStubForHost(
       final String remoteHost) {
     if (perHostClientStubMap.containsKey(remoteHost)) {
-      return perHostClientStubMap.get(remoteHost);
+      return perHostClientStubMap.get(remoteHost).get();
     }
+    return addOrUpdateClientStubForHost(remoteHost);
+  }
 
-    final InterNodeRpcServiceGrpc.InterNodeRpcServiceStub stub = buildStubForHost(remoteHost);
-    perHostClientStubMap.put(remoteHost, stub);
+  private synchronized InterNodeRpcServiceStub addOrUpdateClientStubForHost(final String remoteHost) {
```

Comment:
For a synchronized method, can we add the callers/threads ?

Reply:
Added.